kube-client/src/config/file_config.rs

```diff
@@ -574,14 +587,21 @@ users:
     client-certificate-data: aGVsbG8K
     client-key-data: aGVsbG8K
 "#;
-        let file = tempfile::NamedTempFile::new().expect("create config tempfile");
-        fs::write(file.path(), config_yaml).unwrap();
-        let cfg = Kubeconfig::read_from(file.path())?;
+        let cfg = Kubeconfig::from_yaml(config_yaml)?;
 
         // Ensure we have data from both documents:
         assert_eq!(cfg.clusters[0].name, "k3d-promstack");
         assert_eq!(cfg.clusters[1].name, "k3d-k3s-default");
 
         Ok(())
     }
+
+    #[test]
+    fn kubeconfig_from_empty_string() -> Result<(), KubeconfigError> {
+        let cfg = Kubeconfig::from_yaml("")?;
+
+        assert_eq!(cfg, Kubeconfig::default());
```

Comment:
Do we really want an empty string to default? This is inconsistent with loading from file(s) which errors when there's no config.

https://github.com/kube-rs/kube-rs/blob/3f89ac06b36f620cca79fa75ced38425fa2a95be/kube-client/src/config/file_config.rs#L261

Reply:
Yeah, for the sake of consistency it may be better to fail in that case. If so, what do you think we should do with the error code? `EmptyConfig` takes `PathBuf` as a payload, but with string input there is no file. Ideas?

Reply:
I don't think we should do `PathBuf::from("<text>")`.
Maybe `Option<PathBuf>` and adjust the message? Or add a new error variant.

Reply:
How about we introduce
```rust
enum KubeconfigSource {
    InCluster,
    FIle(PathBuf),
    Yaml,
}
```

and use it as a payload for `EmptyConfig`?

We potentially can use it in more broad fashion with `Kubeconfig`.

Reply:
Personally, I don't think it's that bad that it's inconsistent with file loading. If some of the files are empty in the big `KUBECONFIG` evar, then we still don't give an error. But, we give the error when there's no file that we can use, because it indicates for sure the config is invalid.

The problem here isn't really that `Kubeconfig::default() == Kubeconfig::from_yaml("")`, but that it is possible to construct an invalid config programmatically. 

If we had something like `Kubeconfig::validate` to run as part of loading (or manually for users who do it programmatically), then this problem is minimised. Maybe this is something we should consider adding? (although it sounds like that would be difficult and probably something for a separate issue).

Reply:
Yeah, this is rapidly expanding into something else. I wouldn't mind to tackle this, but is clearly calls for more thorough design and definitely separate issue.
I just pushed quick `KubeconfigSource` implementation. LMK whether you want it in or out.

Reply:
> If some of the files are empty in the big `KUBECONFIG` evar, then we still don't give an error.

Do you mean stacked kubeconfig with `KUBECONFIG=a:b:c`? I think it errors.

> I just pushed quick `KubeconfigSource` implementation.

We have separate errors for in-cluster and file (`kubeconfig`) config. So `KubeconfigSource::Incluster` doesn't make sense. I think it's better to add a new error for this method. It doesn't fit well within the current errors because it's a different source.

Reply:
Maybe just removing `InCluster` is enough. Creating a new error is trickier than I thought.

Reply:
> Do you mean stacked kubeconfig with KUBECONFIG=a:b:c? I think it errors.

Yeah, you are right. It does actually error. It probably doesn't have to error from our perspective. `kubectl` deals with it. It actually replaces the empty file with a defaulted Kubeconfig (but preserving the `current-context` across if it needed to...)

I'm inclined to not have a new enum source here or error for this. If users are making these things from code, I don't think we need to present it as a quasi-validation error because it's entirely in their control what they want to put in there. We also have the case that `kubectl` will default an empty kubeconfig file to our default.

Reply:
Oh, `kubectl` deals with it?

Maybe I misinterpreted this when implementing it:

> - Produce errors for files with content that cannot be deserialized.
>
> https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#merging-kubeconfig-files

I guess "**with** content" means empty file is not an error?

Reply:
I think "with content" here just means yaml. kubectl will actually truncate any file that is valid yaml to ::default, but preserving keys it understands. It produces errors only if you put non-yaml or something with mismatching types.

Reply:
Yeah, looks like we should default to empty `Kubeconfig`. `Kubeconfig::from_yaml("")` shouldn't error.

```bash
touch test && KUBECONFIG=/tmp/test kubectl config view
```
outputs
```yaml
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users: null
```

```bash
echo test > test && KUBECONFIG=/tmp/test kubectl config view
```
```text
error:
error loading config file "/tmp/test":
couldn't get version/kind; json parse error:
json:
cannot unmarshal string into Go value of type struct { APIVersion string "json:\"apiVersion,omitempty\""; Kind string "json:\"kind,omitempty\"" }
```

- empty, `---`, `{}`, `{"foo":1}`: defaults
- mismatching types (`[]`, `test`): errors

Reply:
@imp You can revert the error.

I'll open a separate PR to fix the rest.

Reply:
Right, so I am changing the empty_string test to check for success.  Also in this case it seems like deriving `Default` for `Kubeconfig` is not entirely correct. It should be rather implemented with the values above.

Reply:
> Right, so I am changing the empty_string test to check for success.

Yes please. Sorry about all the time this took. Complexity lurking at all the corners.

> Seems like deriving Default for Kubeconfig is not entirely correct

That is still fine to keep tbh. It's a helper for manually constructing one, and taking it out would be breaking change. We are settling on default == empty here anyway, so it feels good to me.

Reply:
Very well